src/navigation_controller.ts

```diff
@@ -109,96 +107,95 @@ export class NavigationController {
    * Adds all necessary event listeners and markers to a workspace for keyboard
    * navigation to work. This must be called for keyboard navigation to work
    * on a workspace.
-   * @param {!Blockly.WorkspaceSvg} workspace The workspace to add keyboard
+   *
+   * @param workspace The workspace to add keyboard
    *     navigation to.
-   * @public
    */
-  addWorkspace(workspace) {
+  addWorkspace(workspace: WorkspaceSvg) {
     this.navigation.addWorkspace(workspace);
   }
 
   /**
    * Removes all necessary event listeners and markers to a workspace for
    * keyboard navigation to work.
-   * @param {!Blockly.WorkspaceSvg} workspace The workspace to remove keyboard
+   *
+   * @param workspace The workspace to remove keyboard
    *     navigation from.
-   * @public
    */
-  removeWorkspace(workspace) {
+  removeWorkspace(workspace: WorkspaceSvg) {
     this.navigation.removeWorkspace(workspace);
   }
 
   /**
    * Turns on keyboard navigation.
-   * @param {!Blockly.WorkspaceSvg} workspace The workspace to turn on keyboard
+   *
+   * @param workspace The workspace to turn on keyboard
    *     navigation for.
-   * @public
    */
-  enable(workspace) {
+  enable(workspace: WorkspaceSvg) {
     this.navigation.enableKeyboardAccessibility(workspace);
   }
 
   /**
    * Turns off keyboard navigation.
-   * @param {!Blockly.WorkspaceSvg} workspace The workspace to turn off keyboard
+   *
+   * @param workspace The workspace to turn off keyboard
    *     navigation on.
-   * @public
    */
-  disable(workspace) {
+  disable(workspace: WorkspaceSvg) {
     this.navigation.disableKeyboardAccessibility(workspace);
   }
 
   /**
    * Gives the cursor to the field to handle if the cursor is on a field.
-   * @param {!Blockly.WorkspaceSvg} workspace The workspace to check.
-   * @param {!ShortcutRegistry.KeyboardShortcut} shortcut The shortcut
+   *
+   * @param workspace The workspace to check.
+   * @param shortcut The shortcut
    *     to give to the field.
-   * @returns {boolean} True if the shortcut was handled by the field, false
+   * @returns True if the shortcut was handled by the field, false
    *     otherwise.
-   * @protected
    */
-  fieldShortcutHandler(workspace, shortcut) {
+  protected fieldShortcutHandler(
+    workspace: WorkspaceSvg,
+    shortcut: ShortcutRegistry.KeyboardShortcut,
+  ): boolean {
     const cursor = workspace.getCursor();
     if (!cursor || !cursor.getCurNode()) {
       return false;
     }
     const curNode = cursor.getCurNode();
     if (curNode.getType() === ASTNode.types.FIELD) {
-      return /** @type {!Blockly.Field} */ (curNode.getLocation()).onShortcut(
-        shortcut,
-      );
+      return (curNode.getLocation() as Blockly.Field).onShortcut(shortcut);
```

Comment:
In the future we could probably switch this to use a type guard so you don't have to cast. But might be better to just have these be actual subclasses or generics.

Reply:
Yeah, this problem is what made me consider pulling over all of the AST code as well. I think there are typescript-y ways to get the types to behave right on this without casts, and it's a common pattern.